### PR TITLE
Verify checksums wherever possible

### PR DESCRIPTION
**Pull request by dholl**
Here's a PR for issue #28.  I also added a new log level.  You had D for debug, I for info, and E for error, right?  I added N for notice --- which is just one step higher than info.  In this case, I used notice level to produce a bold message during git checksum verification, when there are newer commits than the provided checksum.  Maybe its overkill?  I was interested to get an automatic message when newer commits were present, but I didn't want the message to be drowned out in the rest of the text flying by.  Thus, I made N produce bold text.

### CODE REVIEW

**Comment by dholl** [comment]
Howdy @cfriedt !  Do you have any thoughts or changes I should incorporate to make this PR ready for inclusion?  (Drop the ```N``` and just use I or D for the message about newer git commits?  Or anything else?)

**Comment by dholl** [comment]
oops, I had missed a TMP_DIR in one place.

**Comment by cfriedt** [comment]
Weird, I thought I already commented a while back, but I would definitely not warn if there are newer git commits, particularly because (in some cases) people want to stick to tagged releases.

**Comment by dholl** [comment]
Yeah, I didn't feel the need to label this message as "W"arn either -- hence which I called it "N"otice.  By default, the git checksum verification produces a "D"ebug message for good matches, and an "E"rror message if the git hash wasn't found anywhere in the git history --- meaning that someone either rewrote their entire git branch, or perhaps there's a man-in-the-middle attack going on and an adversary tried giving us a completely different git repo.  (that's the purpose of checking hashes of downloaded files right?)  :)

Additionally, I noticed that build.sh tends to pull from a release or branch for git sources, but doesn't pin to any specific git commit, right?  So the current build.sh behavior silently allows upstream to push additional commits which build.sh would happily pick up.  So in the git checksum verification, I continue allowing this existing behavior, but I just log an "N"otice about it -- in case whomever is building encounters unexpected behavior and would like a clue.  Or if we'd like to really pin to an exact git commit from each source git repo, then we should probably checkout exact hashes, right?

**Comment by cfriedt** [comment]
the BRANCH argument to fetch() can be a branch, a tagged commit (equivalent to a commit) or a regular commit.

File: build.sh [hunk 1]
```diff
@@ -98,19 +103,73 @@ function ncpus() {
...
+  test "$( git -C "${1}" cat-file -t "${2}" 2>/dev/null )" = commit \
+    || E "Repository '${1}' does not match '${2}'"
+  # Then verify the hash is in the current branch.  (The branch may have newer commits.)
+  #    http://stackoverflow.com/questions/4127967/validate-if-commit-exists
+  test "$( git -C "${1}" rev-list HEAD.."${2}" | wc -l )" -eq 0 \
+    || E "Repository '${1}' does not match '${2}'"
+  # Lastly, emit a notice if the repo appears to have newer commits, but do not fail:
+  test "$( git -C "${1}" rev-parse HEAD )" = "${2}" \
+    && D "Repository '${1}' matches '${2}'" \
+    || N "Repository '${1}' has newer commits than '${2}'"
```
Comment (dholl) [inline build.sh]: Done!

**Comment by cfriedt** [comment]
Since there is no real harm in printing out a notice if there are newer commits to a git repository, and since there is zero impact on users, in the interest of moving things along I'll merge it as-is.

**Comment by dholl** [comment]
Ahh, heh, I had just removed the notice in commit 40d2568.  But I would personally recommend to keep the notice, because without it, there is no other notice or protection against an "MitM" attack consisting of a malicious user cloning the source repo, adding their own commits on top of the branch / tag / whatever, and then hijacking the tcp connection when build.sh clone's the repo.  (and build.sh downloads the repo over http instead of https, nor any other DNSSEC or other minimal protections...  --- not that GitHub.com supports DNSSEC anyway, but that's another nuisance...)

If you'd like to re-incorporate the notice reporting newer commits, then just "git revert 40d2568".  :)

**Comment by cfriedt** [comment]
Hmm... why don't we just switch to using https for everything. I think the only reason I used http was because I wasn't aware originally about "curl -L".

Issue #32 created!

**Pull request merged**